File: core/src/main/java/com/graphhopper/routing/util/parsers/OSMCrossingParser.java
package com.graphhopper.routing.util.parsers;

import com.graphhopper.reader.ReaderWay;
import com.graphhopper.routing.ev.Crossing;
import com.graphhopper.routing.ev.EnumEncodedValue;
import com.graphhopper.storage.IntsRef;
import com.graphhopper.util.Helper;

import java.util.List;
import java.util.Map;

/**
 * Parses the node information regarding crossing=* and railway=*
 */
public class OSMCrossingParser implements TagParser {
    protected final EnumEncodedValue<Crossing> crossingEnc;

    public OSMCrossingParser(EnumEncodedValue<Crossing> crossingEnc) {
        this.crossingEnc = crossingEnc;
    }

    @Override
    public void handleWayTags(IntsRef edgeFlags, ReaderWay readerWay, IntsRef relationFlags) {
        List<Map<String, Object>> nodeTags = readerWay.getTag("node_tags", null);
        if (nodeTags == null)
            return;

        for (int i = 0; i < nodeTags.size(); i++) {
            Map<String, Object> tags = nodeTags.get(i);
            if ("crossing".equals(tags.get("railway")) || "level_crossing".equals(tags.get("railway"))) {
                String barrierVal = (String) tags.get("crossing:barrier");
                crossingEnc.setEnum(false, edgeFlags, (Helper.isEmpty(barrierVal) || "no".equals(barrierVal)) ? Crossing.RAILWAY : Crossing.RAILWAY_BARRIER);
                return;
            }

            String crossingSignals = (String) tags.get("crossing:signals");
            if ("yes".equals(crossingSignals)) {
                crossingEnc.setEnum(false, edgeFlags, Crossing.TRAFFIC_SIGNALS);
                return;
            }

            String crossingMarkings = (String) tags.get("crossing:markings");
            if ("yes".equals(crossingMarkings)) {
                crossingEnc.setEnum(false, edgeFlags, Crossing.MARKED);
                return;
            }

            String crossingValue = (String) tags.get("crossing");
            // some crossing values like "no" do not require highway=crossing and sometimes no crossing value exists although highway=crossing
            if (Helper.isEmpty(crossingValue) && ("no".equals(crossingSignals) || "no".equals(crossingMarkings)
                    || "crossing".equals(tags.get("highway")) || "crossing".equals(tags.get("footway")) || "crossing".equals(tags.get("cycleway")))) {
                crossingEnc.setEnum(false, edgeFlags, Crossing.UNMARKED);
                // next node could have more specific Crossing value
                continue;
            }
            Crossing crossing = Crossing.find(crossingValue);
            if (crossing != Crossing.MISSING)
                crossingEnc.setEnum(false, edgeFlags, crossing);
        }
    }
}

File: core/src/main/kotlin/com/graphhopper/routing/util/parsers/OSMCrossingParser.kt
package com.graphhopper.routing.util.parsers

import com.graphhopper.reader.ReaderWay
import com.graphhopper.routing.ev.Crossing
import com.graphhopper.routing.ev.EdgeIntAccess
import com.graphhopper.routing.ev.EnumEncodedValue
import com.graphhopper.storage.IntsRef
import com.graphhopper.util.Helper

/**
 * Parses the node information regarding crossing=* and railway=*
 */
class OSMCrossingParser(private val crossingEnc: EnumEncodedValue<Crossing>) : TagParser {

    override fun handleWayTags(edgeId: Int, edgeIntAccess: EdgeIntAccess, way: ReaderWay, relationFlags: IntsRef?) {
        val nodeTags: List<Map<String, Any>> = way.getTag("node_tags", null) ?: return

        for (i in nodeTags.indices) {
            val tags = nodeTags[i]
            if ("crossing" == tags["railway"] || "level_crossing" == tags["railway"]) {
                val barrierVal = tags["crossing:barrier"] as String?
                crossingEnc.setEnum(false, edgeId, edgeIntAccess,
                    if (Helper.isEmpty(barrierVal) || "no" == barrierVal) Crossing.RAILWAY else Crossing.RAILWAY_BARRIER)
                return
            }

            val crossingSignals = tags["crossing:signals"] as String?
            if ("yes" == crossingSignals) {
                crossingEnc.setEnum(false, edgeId, edgeIntAccess, Crossing.TRAFFIC_SIGNALS)
                return
            }

            val crossingMarkings = tags["crossing:markings"] as String?
            if ("yes" == crossingMarkings) {
                crossingEnc.setEnum(false, edgeId, edgeIntAccess, Crossing.MARKED)
                return
            }

            val crossingValue = tags["crossing"] as String?
            // some crossing values like "no" do not require highway=crossing and sometimes no crossing value exists although highway=crossing
            if (Helper.isEmpty(crossingValue) && ("no" == crossingSignals || "no" == crossingMarkings
                        || "crossing" == tags["highway"] || "crossing" == tags["footway"] || "crossing" == tags["cycleway"])) {
                crossingEnc.setEnum(false, edgeId, edgeIntAccess, Crossing.UNMARKED)
                // next node could have more specific Crossing value
                continue
            }
            val crossing = Crossing.find(crossingValue)
            if (crossing != Crossing.MISSING)
                crossingEnc.setEnum(false, edgeId, edgeIntAccess, crossing)
        }
    }
}
